### DISCUSSION TITLE
Use of PlainObjectBase in lieu of MatrixBase

### DISCUSSION THREAD

**Discussion by Poupine** [libigl/libigl#1920] [General]
Hi,

I've recently used the arap solver implemented in libigl and noticed that the functions expect `Eigen::PlainObjectBase<Derived>` instead of `Eigen::MatrixBase<Derived>` which to me would make more sense. By expecting a PlainObjectBase, the caller is prevented from passing an instance of a Map as input to the function, which makes an extra copy mandatory in case the data already exists in memory but not in an Eigen::Matrix object.
Was the intent with using PlainObjectBase to allow passing either a Matrix or Array as input? In which case wouldn't DenseBase<Derived> be a better choice to allow passing mapped buffers and save a copy to a temp Matrix or Array?

Cheers,

Jérémy

**Comment by alecjacobson** [comment]
Inputs should be MatrixBase and outputs should be PlainObjectBase (to allow
resizing).

A caveat IIRC is that MatrixBase alsot
 allows expressions. This can be really costly if inside the function it's
then accessed (and evaluated) multiple times. In that case .eval should be
called first.

**Reply by Poupine** [reply]
Here are the changes I would make:
`template <typename DerivedV, typename DerivedF, typename Derivedb>
IGL_INLINE bool
arap_precomputation(const Eigen::PlainObjectBase<DerivedV> &V,
                    const Eigen::PlainObjectBase<DerivedF> &F, const int dim,
                    const Eigen::PlainObjectBase<Derivedb> &b, ARAPData &data);`
would become
`template <typename DerivedV, typename DerivedF, typename Derivedb>
IGL_INLINE bool
arap_precomputation(const Eigen::MatrixBase<DerivedV> &V,
                    const Eigen::MatrixBase<DerivedF> &F, const int dim,
                    const Eigen::PlainObjectBase<Derivedb> &b, ARAPData &data);`

Then the implementation in the cpp file needs to be adapted slightly (I have those changes and can create a PR). Then it becomes possible to call `arap_precomputation(Eigen::Map<const MatrixXd>(vertices.data()), Eigen::Map<const MatrixXi>(indices.data()), ...)` which makes it possible (without extra copy) to input data from a custom data structure representing a mesh. 
It offers more flexibility this way because the user of the library is free to implement their mesh structure however they want (as long as the underlying data can be wrapped in an eigen map). What do you think?

MatrixBase does indeed allow expressions which as you pointed out may be costly but I'm thinking that in this case, running a profiler would indicate such behaviour and it would be trivial to fix by calling .eval(). This could be pointed out in the comments maybe?
